src/support/allocators/node_allocator.h

```diff
@@ -0,0 +1,356 @@
+ *   can have an advantageous access pattern which leads to less cache misses.
+ *
+ * ## Design & Implementation
+ *
+ * Allocator is a cheaply copyable, `std::allocator` compatible type used for the containers. Similar to
+ * `std::pmr::polymorphic_allocator`, it holds a pointer to a memory resource.
+ *
+ * MemoryResource is an immobile object that actually allocates, holds and manages chunks of memory. Since there is
+ * unfortunately no way to determine the size of nodes that we want to optimize for in advance, MemoryResource uses
+ * a simple heuristic: We assume the first call to Allocate with 1 element is for the node, and upon that first call the
+ * MemoryResource is configured to use that as it's chunk size. Note that due to this heuristic we cannot implement an
+ * `std::pmr::memory_resource` because it does not gie the information about the number of elements.
+ *
+ * ## Further Links
+ *
+ * @see CppCon 2017: Bob Steagall “How to Write a Custom Allocator” https://www.youtube.com/watch?v=kSWfushlvB8
+ * @see C++Now 2018: Arthur O'Dwyer “An Allocator is a Handle to a Heap” https://www.youtube.com/watch?v=0MdSJsCTRkY
+ * @see AllocatorAwareContainer: Introduction and pitfalls of propagate_on_container_XXX defaults
+ *     https://www.foonathan.net/2015/10/allocatorawarecontainer-propagation-pitfalls/
+ */
+namespace node_allocator {
+
+/**
+ * Actually holds and provides memory to an allocator. MemoryResource is an immobile object. It stores a number of memory blocks
+ * (the pool) which are used to quickly give out memory of a fixed chunk size. The class is purposely kept very simple. It only
+ * knows about "Chunks" and "Blocks".
+ *
+ * - Block: MemoryResource allocates one memory block at a time. These blocks are kept around until the memory resource is destroyed.
+ *
+ * - Chunks: Node-based containers allocate one node at a time. Whenever that happens, the MemoryResource's Allocate() gives out
+ *   one chunk of memory. These chunks are carved out from a previously allocated memory block. Whenever a node is given back
+ *   with Deallocate(), it is put into a freelist.
+ */
+class MemoryResource
+{
+public:
+    /**
+     * Inplace linked list of the allocation chunks, used for the free list.
+     */
+    struct ChunkNode {
+        void* next;
+    };
+
+    /**
+     * Construct a new Memory Resource object and uses the specified block size for allocations.
+     * Actually the real block size can be a bit smaller, it will be the largest multiple of chunk size
+     * that fits into the block.
+     */
+    explicit MemoryResource(size_t block_size_byte) : m_block_size_bytes(block_size_byte) {}
+
+    MemoryResource() = default;
+
+    /**
+     * Copying a memory resource is not allowed, it is an immobile object.
+     */
+    MemoryResource(const MemoryResource&) = delete;
+    MemoryResource& operator=(const MemoryResource&) = delete;
```

Comment:
Perhaps rule of 5 (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-five, https://www.stroustrup.com/C++11FAQ.html#default)
```diff 
     /**
-     * Copying a memory resource is not allowed, it is an immobile object.
+     * Copying/moving a memory resource is not allowed; it is an immobile object.
      */
     MemoryResource(const MemoryResource&) = delete;
     MemoryResource& operator=(const MemoryResource&) = delete;
+    MemoryResource(MemoryResource&&) = delete;
+    MemoryResource& operator=(MemoryResource&&) = delete;
```